lib/msf/core/rhosts_walker.rb

```diff
@@ -0,0 +1,208 @@
+# -*- coding: binary -*-
+
+module Msf
+  ###
+  #
+  # Parses the RHOSTS datastore value, and yields the possible combinations of datastore values
+  # that exist for each host
+  #
+  ###
+  class RhostsWalker
+
+    ###
+    # An error which additionally keeps track of a particular rhost substring which resulted in an error when enumerating
+    # the provided rhost string
+    ###
+    class Error < StandardError
+      attr_reader :value, :cause
+
+      def initialize(value, msg = "Unexpected rhost value: #{value.inspect}", cause: nil)
+        super(msg)
+        @value = value
+        @cause = cause
+        set_backtrace(cause.backtrace) if cause
+      end
+    end
+
+    def initialize(value = '', datastore = Msf::ModuleDataStore.new(nil))
+      @value = value
+      @datastore = datastore
+    end
+
+    #
+    # Iterate over the valid rhosts datastores. This can be combined Calling `#valid?` beforehand to ensure
+    # that there are no invalid configuration values, as they will be ignored by this method.
+    #
+    # @yield [Msf::DataStore] Yields only _valid_ rhost values.
+    def each(&block)
+      return unless @value
+      return unless block_given?
+
+      parse(@value, @datastore).each do |result|
+        block.call(result) if result.is_a?(Msf::DataStore)
+      end
+
+      nil
+    end
+
+    # Count the _valid_ datastore permutations for the current rhosts value. This count will
+    # ignore any invalid values.
+    #
+    # @return [Integer]
+    def count
+      to_enum.count
+    end
+
+    #
+    # Retrieve the list of errors associated with this rhosts walker
+    # @yield [Msf::RhostsWalker::Error] Yields only invalid rhost values.
+    def errors(&block)
+      return unless @value
+      return unless block_given?
+
+      parse(@value, @datastore).each do |result|
+        block.call(result) if result.is_a?(Msf::RhostsWalker::Error)
+      end
+
+      nil
+    end
+
+    #
+    # Indicates that the rhosts value is valid and iterable
+    #
+    # @return [Boolean] True if all items are valid, and there are at least some items present to iterate over. False otherwise.
+    def valid?
+      parsed_values = parse(@value, @datastore)
+      parsed_values.all? { |result| result.is_a?(Msf::DataStore) } && parsed_values.count > 0
+    rescue StandardError => e
+      elog('rhosts walker invalid', error: e)
+      false
+    end
+
+    #
+    # Parses the input rhosts string, and yields the possible combinations of datastore values.
+    #
+    # @param value [String] the http string
+    # @param datastore [Msf::Datastore] the datastore
+    # @return [Enumerable<Msf::DataStore|StandardError>] The calculated datastore values that can be iterated over for
+    #   enumerating the given rhosts, or the error that occurred when iterating over the input
+    def parse(value, datastore)
+      Enumerator.new do |results|
+        values = value.to_s.split(', ').map { |line| line.split(' ') }.flatten
```

Comment:
Leaving a note in case I forget; Are there any edge cases with spaces in URLs/Passwords etc containing spaces or commas?

Reply:
Additionally, if the password has `@` in it:
```
http://user:p4$$w@rd@example.com
```

Likewise commas could appear in a password or query string pretty easily

Reply:
This is that sort of Ruby magic which is massively convenient but will be a nightmare to debug if folks aren't aware of it - i think it would help to write out the intended logic of these operations in comments for posterity.

Reply:
Fixed by moving to `Addressable` which is better at handling the sorts of input our users would want to make use of:
```
2.5.9 :009 > ::Addressable::URI.parse('http://user:p4$$w@rd@example.com').user
 => "user" 
2.5.9 :010 > ::Addressable::URI.parse('http://user:p4$$w@rd@example.com').password
 => "p4$$w@rd" 
```

Versus Ruby's URI, which is closer to the specification, but a lot more arduous for users as it would require them to URI encode their passwords:
```
2.5.9 :011 > URI.parse('http://user:p4$$w@rd@example.com').password
Traceback (most recent call last):
        5: from /Users/user/.rvm/rubies/ruby-2.5.9/bin/irb:11:in `<main>'
        4: from (irb):11
        3: from /Users/user/.rvm/rubies/ruby-2.5.9/lib/ruby/2.5.0/uri/common.rb:237:in `parse'
        2: from /Users/user/.rvm/rubies/ruby-2.5.9/lib/ruby/2.5.0/uri/rfc3986_parser.rb:73:in `parse'
        1: from /Users/user/.rvm/rubies/ruby-2.5.9/lib/ruby/2.5.0/uri/rfc3986_parser.rb:67:in `split'
URI::InvalidURIError (bad URI(is not URI?): http://user:p4$$w@rd@example.com
```